File: 1522-diameter-of-n-ary-tree/1522-diameter-of-n-ary-tree.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    int maxD =0;
    int getMaxDia(Node* node) {
        //Base case
        if(node->children.size()==0) return 0;
        
        int maxH1 = -1;
        int maxH2 = -1;
        
        for(Node* child : node->children) {
            int childH = getMaxDia(child);
            if(childH > maxH1) {
                maxH2 = maxH1;
                maxH1 = childH;
            }
            else if(childH > maxH2) {
                maxH2 = childH;
            }
            
            maxD = max(maxD,maxH1+maxH2+2);
        }
        return maxH1+1;
    }
    int diameter(Node* root) {
        getMaxDia(root);
        return maxD;
    }
};

File: 1522-diameter-of-n-ary-tree/1522-diameter-of-n-ary-tree.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    
    public Node() {
        children = new ArrayList<Node>();
    }
    
    public Node(int _val) {
        val = _val;
        children = new ArrayList<Node>();
    }
    
    public Node(int _val,ArrayList<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public int height(Node node) {
        if(node==null) return 0;
        
        int nodeH = 0;
        for(Node child:node.children) {
            nodeH = Math.max(nodeH,height(child));
        }
        
        return nodeH + 1;
    }
    public int maxH =0;
    public void findMaxDia(Node node) {
        int firstH = 0;
        int secondH = 0;
        for(Node child:node.children) {
            int childH = height(child);
            if(childH>=firstH){
                secondH = firstH;
                firstH = childH;
            }
            else if(childH>=secondH) {
                secondH = childH;
            }
            maxH = Math.max(maxH,firstH+secondH);
        }
        
        for(Node child : node.children) {
            findMaxDia(child);
        }
    }
    public int diameter(Node root) {
        findMaxDia(root);
        return maxH;
    }
}